cnd/src/network/protocol.rs

```diff
@@ -0,0 +1,180 @@
+use crate::swap_protocols::{
+    rfc003::{Secret, SecretHash},
+    SwapId,
+};
+use futures::{AsyncRead, AsyncWrite, Future};
+use libp2p::core::{upgrade, InboundUpgrade, OutboundUpgrade, UpgradeInfo};
+use serde::{Deserialize, Serialize};
+use std::{iter, pin::Pin, str::FromStr};
+use tracing::trace;
+
+// TODO: Do this in a more elegant manner.
+lazy_static::lazy_static! {
+    static ref JSON_INFO_LENGTH: usize = {
+        let swap_id = SwapId::from_str("7f3a105d-ecf2-4cc6-b35c-b4351ac28a34").unwrap();
+
+        let bytes = b"hello world, you are beautiful!!";
+        let secret = Secret::from(*bytes);
+        let secret_hash = secret.hash();
+
+        let info = SecretHashInfo {
+            swap_id,
+            secret_hash,
+        };
+
+        let serialized = serde_json::to_vec(&info).unwrap();
+
+        serialized.len()
+    };
+}
+
+/// The secret hash sharing protocol works in the following way:
+///
+/// - Dialer (Alice) writes the `SecretHashInfo` to the substream.
+/// - Listener (Bob) reads the `SecretHashInfo` from the substream.
+
+/// Represents a prototype for an upgrade to handle the secret hash sharing
+/// protocol.
+#[derive(Debug, Clone)]
+pub struct SecretHashProtocolConfig {
+    info: Option<SecretHashInfo>,
+}
+
+impl SecretHashProtocolConfig {
+    /// Sender sends the secret hash `info` to the peer.
+    pub fn sender(info: SecretHashInfo) -> Self {
+        SecretHashProtocolConfig { info: Some(info) }
+    }
+
+    /// Receiver receives the secret hash from the peer.
+    pub fn receiver() -> Self {
+        SecretHashProtocolConfig { info: None }
+    }
+}
+
+/// Data sent to peer in secret hash protocol.
+#[derive(Clone, Deserialize, Debug, Serialize)]
+pub struct SecretHashInfo {
+    swap_id: SwapId,
```

Comment:
You could make these fixed length byte arrays, then the length is going to be guaranteed at compile-time and you can chuck all these tests out.

Reply:
done, nice.